File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] cand, int target) {
        List<List<Integer>> ans = new ArrayList<>(); 
        List<Integer> temp = new ArrayList<>();  
        int index = 0; 
        help(index,ans,cand,target,temp); 
        return ans;
    }
    static void help(int index,List<List<Integer>>ans,int arr[],int target,List<Integer> temp) {
       
		if(index == arr.length) {
            if(target == 0) {
                ans.add(new ArrayList<>(temp)); 
            }
            return;
        }
		
		
        if(arr[index] <= target) {
            temp.add(arr[index]);
            help(index,ans,arr,target-arr[index],temp);
            temp.remove(temp.size()-1); 
        }
        help(index+1,ans,arr,target,temp); 
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates, target):
        ans = []
        # candidates.sort()
        self.help(candidates, target, 0, [],ans)
        return ans
    
    def help(self, nums, target, index, temp, ans):
        if target < 0:
            return
        if target == 0:
            ans.append(temp)
            return 
        for i in range(index, len(nums)):
            self.help(nums, target-nums[i], i, temp+[nums[i]], ans)